pennylane_lightning/lightning_kokkos.py

```diff
@@ -0,0 +1,817 @@
+            return self.state_vector
+
+        def adjoint_jacobian(self, tape, starting_state=None, use_device_state=False):
+            if self.shots is not None:
+                warn(
+                    "Requested adjoint differentiation to be computed with finite shots."
+                    " The derivative is always exact when using the adjoint "
+                    "differentiation method.",
+                    UserWarning,
+                )
+
+            tape_return_type = self._check_adjdiff_supported_measurements(tape.measurements)
+
+            if not tape_return_type:  # the tape does not have measurements
+                return np.array([], dtype=self.state.dtype)
+
+            if tape_return_type is State:
+                raise QuantumFunctionError(
+                    "This method does not support statevector return type. "
+                    "Use vjp method instead for this purpose."
+                )
+
+            self._check_adjdiff_supported_operations(tape.operations)
+
+            processed_data = self._process_jacobian_tape(tape, starting_state, use_device_state)
+
+            if not processed_data:  # training_params is empty
+                return np.array([], dtype=self.state.dtype)
+
+            trainable_params = processed_data["tp_shift"]
+
+            # If requested batching over observables, chunk into OMP_NUM_THREADS sized chunks.
+            # This will allow use of Lightning with adjoint for large-qubit numbers AND large
+            # numbers of observables, enabling choice between compute time and memory use.
+            requested_threads = int(getenv("OMP_NUM_THREADS", "1"))
+
+            adjoint_jacobian = AdjointJacobianC64() if self.use_csingle else AdjointJacobianC128()
+
+            if self._batch_obs and requested_threads > 1:
+                obs_partitions = _chunk_iterable(
+                    processed_data["obs_serialized"], requested_threads
+                )
+                jac = []
+                for obs_chunk in obs_partitions:
+                    jac_local = adjoint_jacobian(
+                        processed_data["state_vector"],
+                        obs_chunk,
+                        processed_data["ops_serialized"],
+                        trainable_params,
+                    )
+                    jac.extend(jac_local)
+            else:
+                jac = adjoint_jacobian(
+                    processed_data["state_vector"],
+                    processed_data["obs_serialized"],
+                    processed_data["ops_serialized"],
+                    trainable_params,
+                )
+            jac = np.array(jac)
+            jac = jac.reshape(-1, len(trainable_params))
+            jac_r = np.zeros((jac.shape[0], processed_data["all_params"]))
+            jac_r[:, processed_data["record_tp_rows"]] = jac
+            return self._adjoint_jacobian_processing(jac_r) if qml.active_return() else jac_r
```

Comment:
Did we have this in the Kokkos device before, or is it a new addition? I cannot recall.

Reply:
It is similar to the old interface (`jac = adj.adjoint_jacobian(self._kokkos_state, obs_serialized, ops_serialized, tp_shift)`). I mimicked `lightning_qubit` as possible, but let me know if it doesn't make sense for some reason. I know we don't support threading over batches, but we might still accept batches and process them sequentially.

Reply:
This is something I played with a few weeks ago around LK, and I think we want to support this, and only create n concurrent SV copies for batching. With a few small changes this could also be batched with mpi4py/dask/ray too, as mentioned before.